File: app/src/main/java/cl/figonzal/lastquakechile/views/activities/SettingsActivity.java
package cl.figonzal.lastquakechile.views.activities;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.appcompat.widget.Toolbar;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.SeekBarPreference;

import cl.figonzal.lastquakechile.R;
import cl.figonzal.lastquakechile.services.NightModeService;
import cl.figonzal.lastquakechile.services.SharedPrefService;
import cl.figonzal.lastquakechile.services.notifications.QuakesNotification;
import timber.log.Timber;

public class SettingsActivity extends AppCompatActivity {

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.settings_activity);

        //Check modo noche
        new NightModeService(this, this.getLifecycle(), new SharedPrefService(getApplicationContext()), getWindow());

        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.settings_menu, new SettingsFragment())
                .commit();

        Toolbar mToolbar = findViewById(R.id.tool_bar);
        setSupportActionBar(mToolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle(R.string.settings);
        }
    }

    @Override
    public void onBackPressed() {

        Intent intent = new Intent(SettingsActivity.this, MainActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);

        super.onBackPressed();
    }

    //Settings Fragment
    public static class SettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

        private Activity activity;
        private SeekBarPreference seekBarPreference;
        private SharedPrefService sharedPrefService;
        private QuakesNotification quakesNotification;

        @Override
        public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {

            setPreferencesFromResource(R.xml.root_preferences, rootKey);

            if (getActivity() != null) {

                activity = getActivity();
                sharedPrefService = new SharedPrefService(getContext());

                quakesNotification = new QuakesNotification(getContext(), sharedPrefService);
            }
        }

        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);

            //Seek bar
            seekBarPreference = findPreference(activity.getString(R.string.SHARED_PREF_LIST_QUAKE_NUMBER));

            if (seekBarPreference != null) {

                seekBarPreference.setMin(10);
                seekBarPreference.setMax(30);

                String limite = String.valueOf((int) sharedPrefService.getData(activity.getString(R.string.SHARED_PREF_LIST_QUAKE_NUMBER), 0));

                //Setear resumen por defecto
                if (limite.equals("0")) {

                    seekBarPreference.setValue(15);
                    seekBarPreference.setSummary(String.format(activity.getString(R.string.LIST_QUAKE_NUMBER_SUMMARY), 15));

                } else {
                    seekBarPreference.setSummary(String.format(activity.getString(R.string.LIST_QUAKE_NUMBER_SUMMARY), Integer.parseInt(limite)));
                }

            } else {
                Timber.e("Seek bar nulo");
            }
        }

        @Override
        public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {

            /*
             * Preferencia Modo Noche
             * MANUAL
             */
            if (key.equals(activity.getString(R.string.NIGHT_MODE_MANUAL_KEY))) {

                //Si el modo manual esta activado

                boolean manualMode = (boolean) sharedPrefService.getData(activity.getString(R.string.NIGHT_MODE_MANUAL_KEY), false);

                if (manualMode) {

                    AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                    activity.setTheme(R.style.AppTheme);
                    activity.recreate();

                    //Mostrar toast modo manual activado
                    Toast.makeText(activity.getApplicationContext(), getString(R.string.NIGHT_MODE_MANUAL_KEY_TOAST_ON), Toast.LENGTH_LONG).show();

                    //Setear automatico como false si manual esta activado
                    sharedPrefService.saveData(activity.getString(R.string.NIGHT_MODE_AUTO_KEY), false);
                }

                //Si modo manual no esta activado
                else {

                    AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                    activity.setTheme(R.style.AppTheme);
                    activity.recreate();

                    //Mostrar toast modo manual desactivado
                    Toast.makeText(activity.getApplicationContext(), activity.getString(R.string.NIGHT_MODE_MANUAL_KEY_TOAST_OFF), Toast.LENGTH_LONG).show();
                }
            }

            //MODO AUTOMATICO
            //Preferencia modo noche automatico
            else if (key.equals(activity.getString(R.string.NIGHT_MODE_AUTO_KEY))) {

                boolean autoMode = (boolean) sharedPrefService.getData(activity.getString(R.string.NIGHT_MODE_MANUAL_KEY), false);

                //Si automatico esta activado, preguntar el estado del modo
                if (autoMode) {

                    AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);

                    activity.setTheme(R.style.AppTheme);
                    activity.recreate();

                    //Mostrar toast con mensaje de modo noche automatico activado
                    Toast.makeText(activity.getApplicationContext(), activity.getString(R.string.NIGHT_MODE_AUTO_KEY_TOAST_ON), Toast.LENGTH_LONG).show();

                    //Setear automatico como false si manual esta activado
                    sharedPrefService.getData(activity.getString(R.string.NIGHT_MODE_MANUAL_KEY), false);
                }

                //Si auto esta desactivado, tema claro por defecto
                else {

                    AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                    activity.setTheme(R.style.AppTheme);
                    activity.recreate();

                    //Mostrar toast con mensaje de modo noche automatico desactivado
                    Toast.makeText(activity.getApplicationContext(), activity.getString(R.string.NIGHT_MODE_AUTO_KEY_TOAST_OFF), Toast.LENGTH_LONG).show();
                }
            }


            /*
             * Preferencias de alertas
             */
            if (key.equals(activity.getString(R.string.FIREBASE_PREF_KEY))) {

                boolean mSuscrito = quakesNotification.checkSuscriptionQuakes();

                //Si el switch esta ON, lanzar toast con SUSCRITO
                if (mSuscrito) {
                    Toast.makeText(getContext(), activity.getString(R.string.FIREBASE_PREF_KEY_TOAST_ALERTAS_ON), Toast.LENGTH_LONG).show();
                }
                //Si el switch esta off lanzar toast con NO SUSCRITO
                else {
                    Toast.makeText(getContext(), activity.getString(R.string.FIREBASE_PREF_KEY_TOAST_ALERTAS_OFF), Toast.LENGTH_LONG).show();
                }

            }

            /*
             * Preferencias de numero de sismos
             */
            if (key.equals(activity.getString(R.string.SHARED_PREF_LIST_QUAKE_NUMBER))) {

                seekBarPreference.setSummary(String.format(activity.getString(R.string.LIST_QUAKE_NUMBER_SUMMARY), seekBarPreference.getValue()));
                sharedPrefService.saveData(activity.getString(R.string.SHARED_PREF_LIST_QUAKE_NUMBER), seekBarPreference.getValue());

                Timber.tag(activity.getString(R.string.TAG_FRAGMENT_SETTINGS)).i(String.format(activity.getString(R.string.TAG_FRAGMENT_SETTINGS_QUAKE_LIST_LIMIT), seekBarPreference.getValue()));
            }

        }

        @Override
        public void onResume() {
            super.onResume();
            getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
        }

        @Override
        public void onPause() {
            super.onPause();
            getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        }
    }
}

File: app/src/main/java/cl/figonzal/lastquakechile/views/activities/SettingsActivity.kt
package cl.figonzal.lastquakechile.views.activities

import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.SeekBarPreference
import cl.figonzal.lastquakechile.R
import cl.figonzal.lastquakechile.core.utils.SharedPrefUtil
import cl.figonzal.lastquakechile.databinding.SettingsActivityBinding
import cl.figonzal.lastquakechile.services.NightModeService
import cl.figonzal.lastquakechile.services.notifications.QuakesNotification
import timber.log.Timber

class SettingsActivity : AppCompatActivity() {

    private lateinit var binding: SettingsActivityBinding

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = SettingsActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)

        //Check modo noche
        NightModeService(this, this.lifecycle, window)

        supportFragmentManager
            .beginTransaction()
            .replace(R.id.settings_menu, SettingsFragment())
            .commit()

        setSupportActionBar(binding.toolbarNoMain.toolBar)

        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setTitle(R.string.settings)
        }


    }

    override fun onBackPressed() {
        val intent = Intent(this@SettingsActivity, MainActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        startActivity(intent)
        super.onBackPressed()
    }

    //Settings Fragment
    class SettingsFragment : PreferenceFragmentCompat(), OnSharedPreferenceChangeListener {

        private var seekBarPreference: SeekBarPreference? = null
        private var sharedPrefUtil: SharedPrefUtil? = null
        private var quakesNotification: QuakesNotification? = null

        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            setPreferencesFromResource(R.xml.root_preferences, rootKey)

            sharedPrefUtil = SharedPrefUtil(requireContext())
            quakesNotification = QuakesNotification(context, sharedPrefUtil)
        }

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)

            //Seek bar
            seekBarPreference =
                findPreference(requireContext().getString(R.string.SHARED_PREF_LIST_QUAKE_NUMBER))

            if (seekBarPreference != null) {

                seekBarPreference!!.min = 10
                seekBarPreference!!.max = 30

                val limite = sharedPrefUtil!!.getData(
                    requireContext().getString(R.string.SHARED_PREF_LIST_QUAKE_NUMBER), 0
                ) as String

                //Setear resumen por defecto
                when (limite) {
                    "0" -> {
                        seekBarPreference!!.value = 15
                        seekBarPreference!!.summary =
                            String.format(
                                requireContext().getString(R.string.LIST_QUAKE_NUMBER_SUMMARY),
                                15
                            )
                    }
                    else -> seekBarPreference!!.summary = String.format(
                        requireContext().getString(R.string.LIST_QUAKE_NUMBER_SUMMARY),
                        limite.toInt()
                    )
                }
            } else {
                Timber.e("Seek bar nulo")
            }
        }

        override fun onSharedPreferenceChanged(preferences: SharedPreferences, key: String) {


            /*
             * Preferencia Modo Noche
             * MANUAL
             */
            if (key == requireContext().getString(R.string.NIGHT_MODE_MANUAL_KEY)) {

                //Si el modo manual esta activado
                when {
                    preferences.getBoolean(
                        requireContext().getString(R.string.NIGHT_MODE_MANUAL_KEY), false
                    ) -> {
                        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
                        requireContext().setTheme(R.style.AppTheme)
                        requireActivity().recreate()

                        //Mostrar toast modo manual activado
                        Toast.makeText(
                            requireContext().applicationContext,
                            getString(R.string.NIGHT_MODE_MANUAL_KEY_TOAST_ON),
                            Toast.LENGTH_LONG
                        ).show()

                        //Setear automatico como false si manual esta activado
                        val edit = preferences.edit()
                        edit.putBoolean(
                            requireContext().getString(R.string.NIGHT_MODE_AUTO_KEY),
                            false
                        )
                        edit.apply()

                    }
                    else -> {
                        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
                        requireContext().setTheme(R.style.AppTheme)
                        requireActivity().recreate()

                        //Mostrar toast modo manual desactivado
                        Toast.makeText(
                            requireContext().applicationContext,
                            requireContext().getString(R.string.NIGHT_MODE_MANUAL_KEY_TOAST_OFF),
                            Toast.LENGTH_LONG
                        ).show()
                    }
                }


            } else if (key == requireContext().getString(R.string.NIGHT_MODE_AUTO_KEY)) {

                //Si automatico esta activado, preguntar el estado del modo
                when {
                    preferences.getBoolean(
                        requireContext().getString(R.string.NIGHT_MODE_AUTO_KEY),
                        false
                    ) -> {
                        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
                        requireContext().setTheme(R.style.AppTheme)
                        requireActivity().recreate()

                        //Mostrar toast con mensaje de modo noche automatico activado
                        Toast.makeText(
                            requireContext().applicationContext,
                            requireContext().getString(R.string.NIGHT_MODE_AUTO_KEY_TOAST_ON),
                            Toast.LENGTH_LONG
                        ).show()

                        //Setear automatico como false si manual esta activado
                        val edit = preferences.edit()
                        edit.putBoolean(
                            requireContext().getString(R.string.NIGHT_MODE_MANUAL_KEY),
                            false
                        )
                        edit.apply()
                    }
                    else -> {
                        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
                        requireContext().setTheme(R.style.AppTheme)
                        requireActivity().recreate()

                        //Mostrar toast con mensaje de modo noche automatico desactivado
                        Toast.makeText(
                            requireContext().applicationContext,
                            requireContext().getString(R.string.NIGHT_MODE_AUTO_KEY_TOAST_OFF),
                            Toast.LENGTH_LONG
                        ).show()
                    }
                }
            }


            /*
             * Preferencias de alertas
             */
            if (key == requireContext().getString(R.string.FIREBASE_PREF_KEY)) {

                val notificationSwitch =
                    preferences.getBoolean(
                        requireContext().getString(R.string.FIREBASE_PREF_KEY),
                        true
                    )

                //Si el switch esta ON, lanzar toast con SUSCRITO
                if (notificationSwitch) {
                    quakesNotification!!.suscribedToQuakes(true)
                    Toast.makeText(
                        context,
                        requireContext().getString(R.string.FIREBASE_PREF_KEY_TOAST_ALERTAS_ON),
                        Toast.LENGTH_LONG
                    ).show()
                } else {
                    quakesNotification!!.suscribedToQuakes(false)
                    Toast.makeText(
                        context,
                        requireContext().getString(R.string.FIREBASE_PREF_KEY_TOAST_ALERTAS_OFF),
                        Toast.LENGTH_LONG
                    ).show()
                }
            }

            /*
             * Preferencias de numero de sismos
             */if (key == requireContext().getString(R.string.SHARED_PREF_LIST_QUAKE_NUMBER)) {

                seekBarPreference!!.summary = String.format(
                    requireContext().getString(R.string.LIST_QUAKE_NUMBER_SUMMARY),
                    seekBarPreference!!.value
                )
                sharedPrefUtil!!.saveData(
                    requireContext().getString(R.string.SHARED_PREF_LIST_QUAKE_NUMBER),
                    seekBarPreference!!.value
                )
                Timber.i(
                    requireContext().getString(R.string.TAG_FRAGMENT_SETTINGS_QUAKE_LIST_LIMIT),
                    seekBarPreference!!.value
                )
            }
        }

        override fun onResume() {
            super.onResume()
            preferenceScreen.sharedPreferences?.registerOnSharedPreferenceChangeListener(this)
        }

        override fun onPause() {
            super.onPause()
            preferenceScreen.sharedPreferences?.unregisterOnSharedPreferenceChangeListener(this)
        }
    }
}